File: Sources/AsyncImageCache/CachedImage.swift
// Sources/AsyncImageCache/CachedImage.swift
//
// A SwiftUI image view backed by the ImageStore cache. Named for the CACHE (not asynchronicity) to avoid
// confusion with SwiftUI's AsyncImage and app-level AsyncImage views: the point of this view is that the
// bytes come from ImageStore's memory + disk cache, decoded/scaled OFF the main thread.
//
// Layout is reserved UP FRONT from an intrinsic size hint (or, once loaded, the cache's known natural size),
// so an image whose dimensions are known - e.g. a chat attachment with server-provided width/height - never
// reflows the transcript when its pixels arrive. While loading, a placeholder fills the reserved box.

import SwiftUI

#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif

public struct CachedImage: View {
    private let url: URL?
    private let intrinsicSize: CGSize?
    private let cornerRadius: CGFloat
    private let contentMode: ContentMode
    private let maxPixelWidth: CGFloat?
    private let store: ImageStore

    @State private var image: PlatformImage?
    // The url `image` and `resolvedSize` belong to. When the view is reused with a NEW url, the old bitmap
    // must not keep displaying (or sizing the box) while the new one loads - body gates on loadedURL == url.
    @State private var loadedURL: URL?
    @State private var resolvedSize: CGSize?

    /// - Parameters:
    ///   - url: the image source (nil renders just the placeholder).
    ///   - intrinsicSize: the source's natural size, if known ahead of load (e.g. from server metadata). When
    ///     provided the view reserves that aspect immediately, so the picture hydrates with ZERO reflow.
    ///   - cornerRadius: rounded-corner clip (a GPU clip, so it costs no CPU mask).
    ///   - contentMode: how the image fills the reserved box (default `.fill`).
    ///   - maxPixelWidth: cap the DECODED width in pixels (the image is downscaled off-main to this width) to
    ///     bound memory for large sources; nil keeps the natural resolution.
    ///   - store: the backing cache (default `.shared`).
    public init(url: URL?,
                intrinsicSize: CGSize? = nil,
                cornerRadius: CGFloat = 0,
                contentMode: ContentMode = .fill,
                maxPixelWidth: CGFloat? = nil,
                store: ImageStore = .shared) {
        self.url = url
        self.intrinsicSize = intrinsicSize
        self.cornerRadius = cornerRadius
        self.contentMode = contentMode
        self.maxPixelWidth = maxPixelWidth
        self.store = store
    }

    // The aspect used to reserve layout: the caller's hint, else the cache's known natural size, else a
    // neutral 4:3 until we learn better.
    private var aspectRatio: CGFloat {
        // Prefer the caller's hint; else the size resolved after loading; else the cache's known natural size
        // - a cheap sync read that survives relaunch via the on-disk xattr - so a cached image reserves its
        // exact box up front with no reflow on hydration. Neutral 4:3 only until anything is known.
        let size = intrinsicSize
            ?? (loadedURL == url ? resolvedSize : nil)
            ?? url.flatMap { store.cachedPixelSize(for: $0) }
        if let size, size.width > 0, size.height > 0 {
            return size.width / size.height
        }
        return 4.0 / 3.0
    }

    public var body: some View {
        Color.clear
            .aspectRatio(aspectRatio, contentMode: .fit)
            .overlay {
                if let image, loadedURL == url {
                    imageView(image)
                        .resizable()
                        .aspectRatio(contentMode: contentMode)
                } else {
                    placeholder
                }
            }
            .clipShape(RoundedRectangle(cornerRadius: cornerRadius, style: .continuous))
            .task(id: url) {
                await load()
            }
    }

    // While loading, fill the reserved box with the image's placeholder grid - a cheap sync read that survives
    // relaunch via the on-disk xattr - upscaled with interpolation into a soft gradient preview of the image,
    // instead of flat gray. Fall back to neutral gray when no placeholder is known (not loaded yet).
    @ViewBuilder
    private var placeholder: some View {
        if let url, let grid = store.placeholder(for: url), let image = ImageProcessing.gridImage(from: grid) {
            // Stretch the NxN grid to fill the box: it was sampled by squishing the whole image into NxN, so
            // stretching back to the box's aspect reconstitutes the proportions. Bilinear = soft gradient.
            imageView(image)
                .resizable()
                .interpolation(.high)
        } else {
            Rectangle().fill(Color.gray.opacity(0.12))
        }
    }

    private func imageView(_ image: PlatformImage) -> Image {
        #if canImport(UIKit)
        return Image(uiImage: image)
        #else
        return Image(nsImage: image)
        #endif
    }

    @MainActor
    private func load() async {
        guard let url else {
            image = nil
            loadedURL = nil
            resolvedSize = nil
            return
        }
        let request = ImageRequest(url: url, targetWidth: maxPixelWidth, cornerRadius: 0)
        if let cached = store.cachedImage(for: request) {
            image = cached
            loadedURL = url
            resolvedSize = store.cachedPixelSize(for: url)
            return
        }
        let loaded = await withCheckedContinuation { continuation in
            store.load(request) { continuation.resume(returning: $0) }
        }
        // task(id: url) cancelled this task if the url changed (or the view went away) while we awaited; a
        // cancelled task's result must not clobber state that now belongs to the newer url. (`self` here is a
        // value copy captured when the task started, so comparing self.url would ALWAYS pass - cancellation
        // is the real staleness check.)
        guard !Task.isCancelled else {
            return
        }
        image = loaded
        loadedURL = url
        resolvedSize = store.cachedPixelSize(for: url)
    }
}

File: android/asyncimagecache/src/main/kotlin/com/abracode/asyncimagecache/CachedImage.kt
// CachedImage.kt
//
// The Compose entry point, a 1:1 port of the SwiftUI `CachedImage` view. It reserves the layout box at the
// correct aspect BEFORE the pixels arrive (zero reflow on hydration), shows the cached blurry placeholder grid
// while loading, then swaps in the ready-to-draw variant - no spinner, no transition, exactly like the Swift
// view.
//
// Faithful to the Swift reference (Sources/AsyncImageCache/CachedImage.swift), NOT the resolved-width sketch in
// porting-plan section 3:
//   - the decode target width is `maxPixelWidth` (the explicit cap), so an uncapped image decodes at its
//     natural resolution, exactly as Swift builds `ImageRequest(url:, targetWidth: maxPixelWidth, cornerRadius: 0)`;
//   - the composable's own `cornerRadius` is a GPU clip (Modifier.clip), not baked into the variant - matching
//     Swift's `.clipShape(RoundedRectangle(...))`. Baking is still available for consumers (the ActionUI element)
//     through `ImageRequest.cornerRadius`.
//
// Aspect-box reservation chain (identical to Swift): explicit `intrinsicSize` -> the size resolved after load ->
// the store's cachedPixelSize -> a neutral 4:3. `resolvedSize`/`loadedUrl` state make the box reflow to the
// image's true aspect once loaded, so an image without a server-provided size still ends up correctly shaped.
//
// Recomposition safety: the load is keyed (LaunchedEffect) on url, so recompositions do not re-request, and a
// changed url cancels the previous LaunchedEffect - its suspended loadAsync resumes into a dead continuation, so
// a stale result can never overwrite the new one (the stale-completion guard Swift gets from task identity). The
// rendered image is additionally gated on `loadedUrl == url` so a reused view never shows the old bitmap.

package com.abracode.asyncimagecache

import android.graphics.Bitmap
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.FilterQuality
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/** How the image fills its reserved box - the Swift `contentMode`. */
enum class ContentMode { Fit, Fill }

private val NEUTRAL_PLACEHOLDER = Color.Gray.copy(alpha = 0.12f)
private const val DEFAULT_ASPECT = 4f / 3f

@Composable
fun CachedImage(
    url: String?,
    modifier: Modifier = Modifier,
    intrinsicSize: Size? = null,
    cornerRadius: Dp = 0.dp,
    contentMode: ContentMode = ContentMode.Fill,
    maxPixelWidth: Float? = null,
    store: ImageStore = ImageStore.shared(LocalContext.current),
    contentDescription: String? = null,
) {
    var image by remember { mutableStateOf<Bitmap?>(null) }
    // The url `image`/`resolvedSize` belong to. When the view is reused with a NEW url, the old bitmap must not
    // keep displaying (or sizing the box) while the new one loads - render + aspect gate on loadedUrl == url.
    var loadedUrl by remember { mutableStateOf<String?>(null) }
    var resolvedSize by remember { mutableStateOf<Size?>(null) }

    // Recomputed on every recomposition (it reads the resolvedSize/loadedUrl state), so the box reflows once the
    // image loads. Precedence identical to Swift: caller hint -> resolved-after-load -> cache's known size -> 4:3.
    val aspectRatio = run {
        val size = intrinsicSize
            ?: (if (loadedUrl == url) resolvedSize else null)
            ?: url?.let { store.cachedPixelSize(it)?.toComposeSize() }
        if (size != null && size.width > 0f && size.height > 0f) size.width / size.height else DEFAULT_ASPECT
    }

    LaunchedEffect(url) {
        if (url == null) {
            image = null; loadedUrl = null; resolvedSize = null
            return@LaunchedEffect
        }
        val request = ImageRequest(url, targetWidth = maxPixelWidth, cornerRadius = 0f)
        store.cachedImage(request)?.let {
            image = it; loadedUrl = url; resolvedSize = store.cachedPixelSize(url)?.toComposeSize()
            return@LaunchedEffect
        }
        // If url changes, this coroutine is cancelled at the suspension point, so the assignments below never run
        // with a stale result.
        val loaded = store.loadAsync(request)
        image = loaded
        loadedUrl = url
        resolvedSize = store.cachedPixelSize(url)?.toComposeSize()
    }

    val contentScale = if (contentMode == ContentMode.Fill) ContentScale.Crop else ContentScale.Fit
    Box(
        modifier = modifier
            .aspectRatio(aspectRatio)
            .clip(RoundedCornerShape(cornerRadius)),
    ) {
        val current = image
        if (current != null && loadedUrl == url) {
            Image(
                bitmap = current.asImageBitmap(),
                contentDescription = contentDescription,
                modifier = Modifier.fillMaxSize(),
                contentScale = contentScale,
            )
        } else {
            // Placeholder grid (soft gradient) while loading, or a neutral fill if none is known yet. Drawn with
            // low filter quality so the 6x6 grid upscales into a smooth blur (the Swift interpolated upscale).
            val preview = url?.let { store.placeholderImage(it) }
            if (preview != null) {
                Image(
                    bitmap = preview.asImageBitmap(),
                    contentDescription = contentDescription,
                    modifier = Modifier.fillMaxSize(),
                    contentScale = contentScale,
                    filterQuality = FilterQuality.Low,
                )
            } else {
                Box(Modifier.fillMaxSize().background(NEUTRAL_PLACEHOLDER))
            }
        }
    }
}

private fun android.util.Size.toComposeSize(): Size = Size(width.toFloat(), height.toFloat())
